[Quant][fx][OLD] Refactor lowering code (part 2)

andrewor14 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #74362

**Summary**: This commit is part 2 of the effort to refactor the
lowering code in _lower_to_native_backend.py. The main change
included in this commit is generalizing the pattern matching
code across different lowering functions. There should be no
change in behavior with this PR.

A future commit will further merge the static and dynamic
lowering code paths.

**Test Plan:**
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

**Reviewers:** jerryzh168

**Subscribers:** jerryzh168

Differential Revision: [D35048525](https://our.internmc.facebook.com/intern/diff/D35048525)

andrewor14 committed:
[Quant][fx] Refactor lowering code (part 2)

Summary: This commit is part 2 of the effort to refactor the
lowering code in _lower_to_native_backend.py. The main changes
included in this commit are:

(1) Merge pattern matching across different lowering functions
(2) Merge the static and dynamic lowering code paths (TODO)

There should be no change in behavior with this PR.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168

Subscribers: jerryzh168

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -360,6 +359,77 @@ def load_arg(a):
+        if ref_node.op != "call_module":
+            return skip_lowering_value
+        ref_class = type(_get_module(ref_node, modules))
+        if ref_class not in STATIC_LOWER_MODULE_MAP and ref_class not in STATIC_LOWER_FUSED_MODULE_MAP:
```
just for consistency can we pass in these maps as well?

jerryzh168 replied:
looks like here we can just combine these two dict into one and do the check together, although to be precise we also need to get the inner class when it is a fused module

jerryzh168 replied:
maybe we can separate these two maps as two arguments since the check required is different

jerryzh168 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -713,23 +748,11 @@ def _lower_quantized_binary_op(
-            continue
-        if bop_node.op != "call_function" or bop_node.target not in binary_ops_to_lower:
+        (q_node, relu_node, bop_node) = _match_static_pattern(
+            n, modules, qconfig_map, binary_ops_to_lower, skip_dequantize_matching=True)
```
why do we skip matching dequantize?

andrewor14 replied:
Because we also allow these patterns:
```
add(dequantize(x), y)
add(x, dequantize(y))
```
Since it's logic specific to binary ops I feel it's better to check dequantize node below (L761) instead of in a common function

jerryzh168 replied:
I see, can we try to match dequantize if it is a Node? in the add scalar cases the non dequantized value would be a constant I think

andrewor14 replied:
Ok, let me try

jerryzh168 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -360,6 +359,77 @@ def load_arg(a):
+    if match_module:
+        if ref_node.op != "call_module":
+            return skip_lowering_value
+        ref_class = type(_get_module(ref_node, modules))
```
module = modules[str(node.target)]

get_target_type(object):
# object can be nn.Module instance or functional/torch op
if isistance(object, nn.Module): return type(object0)
else: return object

andrewor14 committed:
Update on "[Quant][fx] Refactor lowering code (part 2)"


**Summary**: This commit is part 2 of the effort to refactor the
lowering code in _lower_to_native_backend.py. The main change
included in this commit is generalizing the pattern matching
code across different lowering functions. There should be no
change in behavior with this PR.

A future commit will further merge the static and dynamic
lowering code paths.

**Test Plan:**
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

**Reviewers:** jerryzh168

**Subscribers:** jerryzh168

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Refactor lowering code (part 2)"


**Summary**: This commit is part 2 of the effort to refactor the
lowering code in _lower_to_native_backend.py. The main change
included in this commit is generalizing the pattern matching
code across different lowering functions. There should be no
change in behavior with this PR.

A future commit will further merge the static and dynamic
lowering code paths.

**Test Plan:**
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

**Reviewers:** jerryzh168

**Subscribers:** jerryzh168

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Refactor lowering code (part 2)"


**Summary**: This commit is part 2 of the effort to refactor the
lowering code in _lower_to_native_backend.py. The main change
included in this commit is generalizing the pattern matching
code across different lowering functions. There should be no
change in behavior with this PR.

A future commit will further merge the static and dynamic
lowering code paths.

**Test Plan:**
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

**Reviewers:** jerryzh168

**Subscribers:** jerryzh168

Differential Revision: [D35048525](https://our.internmc.facebook.com/intern/diff/D35048525)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35048525).

jerryzh168 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -402,19 +467,20 @@ def _lower_static_weighted_ref_module(
         if ref_class in STATIC_LOWER_FUSED_MODULE_MAP:
             inner_ref_class, q_class = STATIC_LOWER_FUSED_MODULE_MAP[ref_class]
-            if type(ref_module[0]) != inner_ref_class:
+            if type(ref_module[0]) != inner_ref_class:  # type: ignore[index]
                 continue
```
nit: this is also a matching step, I think we can move this closer to the match_static_pattern call or support this inside that function

jerryzh168 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -360,6 +360,83 @@ def load_arg(a):
+      matching_modules_or_ops: either a list of functions or a list of `torch.nn.Module`s.
+          If the reference node is not in this list, then return no match.
+    """
+    skip_lowering_value = (None, None, None)
```
nit: capitalize since this is a constant

jerryzh168 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -360,6 +360,83 @@ def load_arg(a):
+    # (1) All `torch.fx.Node`s must be a dequantize node or a get_attr node (for bias)
+    # (2) There must be at least one dequantize node
+    matched_dequantize = False
+    for arg in ref_node.args:
```
can we pass in the indices of args to match, and then make sure there is at least one match?

jerryzh168 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -360,6 +360,83 @@ def load_arg(a):
+    for arg in ref_node.args:
+        if is_dequantize_node(arg):
+            matched_dequantize = True
+        elif isinstance(arg, Node) and arg.op != "get_attr":
```
this condition feels a bit hacky

jerryzh168 reviewed (approved):
looks great, thanks for the refactoring! had a few comments inline

andrewor14 committed:
Update on "[Quant][fx] Refactor lowering code (part 2)"


**Summary**: This commit is part 2 of the effort to refactor the
lowering code in _lower_to_native_backend.py. The main change
included in this commit is generalizing the pattern matching
code across different lowering functions. There should be no
change in behavior with this PR.

A future commit will further merge the static and dynamic
lowering code paths.

**Test Plan:**
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

**Reviewers:** jerryzh168

**Subscribers:** jerryzh168

Differential Revision: [D35048525](https://our.internmc.facebook.com/intern/diff/D35048525)

[ghstack-poisoned]
